### DISCUSSION TITLE
Using Merge Queue to avoid accidental breakage when queueing multiple PRs.

### DISCUSSION THREAD

**Discussion by storopoli** [rust-bitcoin/rust-bitcoin#3772] [Ideas]
Recently we had an issue in #3758 that unfortunately a high number of queued PRs broke CI on `master` affecting all the ongoing PRs that were targeting it. 

I've commented on that PR that one solution would be to use [GitHub's Merge Queue](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/managing-a-merge-queue).
Since @apoelstra emoji reacted with "taking a look" eyes emoji, I think that opening a discussion makes sense.

Note that we are eligible for merge queue, since we are a public repository hosted by an organization.

## What are Merge Queues

Merge Queues help maintainers safely merge multiple PRs by:

1. Creating a temporary staging branch that combines:

   - The base branch (e.g., `master`)
   - The current PR being tested
   - All PRs ahead in the queue

2. Running required checks (tests, CI) on this combined state before actual merging

This solves two critical problems:

1. **Race Conditions**: Without a queue, multiple PRs merged quickly can break `master` even if they passed tests individually
2. **Time Savings**: Instead of waiting for each PR to be merged and tested sequentially, the queue tests them in batched combinations

Example:
- `PR #1` and `PR #2` both pass tests individually
- But when merged together they could conflict or break functionality
- Merge Queue catches this by testing them together before actual merging

This provides higher confidence that `master` remains stable even with frequent merges from multiple contributors.

## Do Merge Queues break the commit history?

No, Merge Queues don't break or alter the commit/merge history. The temporary staging branches used for testing are discarded after the checks pass.

The final merge to `master` maintains:

1. The original commit history from each PR
2. The proper merge commit (if using merge commits)
3. Linear history (if using squash or rebase merging)

The only difference in the process is:

- PRs are tested together before merging
- Merges happen in a controlled sequence
- The actual merge operation remains the same as configured in your repository (merge commit, squash, or rebase)

So the resulting Git history looks exactly the same as if you had merged the PRs manually - just with additional safety guarantees that the combined changes won't break the codebase.

The staging branches used by the Merge Queue are purely temporary testing environments and don't appear in the final repository history.

**Comment by apoelstra** [comment]
Locally I already have a "merge queue" which fits into my CI pipeline. It interleaves PR checks (which test every commit and ACK or comment on success) and merges (which test the merge commit and then I interactively sign the result and then it pushes the merge). It generally runs operations in the order I queue them except that I have an "afk" mode I can set which defers any interactive signatures until there's nothing else that it can do.

Of course, the limitation here is that this is all running locally on my system, and if e.g. there is a merge conflict, then nobody knows until I post about it. There's also the limitation that it's slow -- because I'm running a bevy of tests on every commit, it may post an "ACK" many hours after I've reviewed the PR (and maybe redundantly, if there has been a force-push since my review). This often leads to a 24-hour round trip where I start something, go to bed, the CI server posts the result, and then Tobin replies.

It seems like using a GH merge queue would improve the first issue (giving public visibility into what's queued to be merged) but wouldn't improve the second (long delays).

Furthermore, I'm very suspicious of the claimed "time savings" in the description of GH merge queues. If this is "saving us time" by testing fewer things we don't want it. It's also not clear to me what happens in the example where PRs 1 and 2 both pass tests but can't be merged together. Will GH merge only PR 1 (as I do) or will it merge neither? If it merges neither that's probably a big regression in terms of PR latency. The other obvious question is how "the actual merge operation remains the same" when our actual merge operation involves cryptographic signatures by maintainers. Is it simply that any push operations to master are queued up and only accepted after a delay? Seems like this would complicate my ability to create merge commits locally.

In general I'm open to trying this but I don't know the answer to these questions.

**Comment by dpc** [comment]
Just enable it.

**Reply by apoelstra** [reply]
AFAICT if we enable it then Github will make its own merge commits, which is a complete non-starter for this repo.

**Reply by dpc** [reply]
AFAIK Merge Queue does not change a single thing about how commits are merged. It purely serializes them and tests one by one before each PR is actually landed, ensuring they are not broken.



@storopoli 

> PRs are tested together before merging

Last time I checked, (maybe a year ago), Github Merge Queue did not support anything w.r.t merging PRs together (unlike e.g. Bors did). It was very conter-intuitive how it worked. Regrettably in our case, because that's something we actually wanted in Fedimint (test togeter and merge together, to save time). Setting parallelism to more than 1 just kind of wastes resources, AFAIK, as each PR will be tested again individually against `master` before actually merging. I mean - it might find the conflict earlier this way, but that's such a rare occurrence, that it seems entirely not worth running test suite multiple times. If that changed in the meantime, I would love to know.

**Reply by dpc** [reply]
In Fedimint we've been rolling with it for about a year with parallelism set to 1.

Serialization of merging is nice because it does prevent from time an accidental breakage of `master`. Also in the CI, we conditionally run much lighter version of our test suite (just to sanity check) at the PR level, and then in the MQ we throw bunch of extra stuff that is slower. This time a contributor and reviewers gets a much faster initial feedback, while the things like e2e flakiness is tested more thoroughly before things reach `master` branch. At the time PR was already reviewed and accepted the extra time does not matter all that much anymore.

**Reply by storopoli** [reply]
> @storopoli
> 
> > PRs are tested together before merging
> 
> Last time I checked, (maybe a year ago), Github Merge Queue did not support anything w.r.t merging PRs together (unlike e.g. Bors did). It was very conter-intuitive how it worked. Regrettably in our case, because that's something we actually wanted in Fedimint (test togeter and merge together, to save time). Setting parallelism to more than 1 just kind of wastes resources, AFAIK, as each PR will be tested again individually against `master` before actually merging. I mean - it might find the conflict earlier this way, but that's such a rare occurrence, that it seems entirely not worth running test suite multiple times. If that changed in the meantime, I would love to know.

I think that's the purpose of the "parallelism" parameter.

**Reply by dpc** [reply]
> No, the GitHub Merge Queue does not combine multiple pull requests (PRs) into a single batch before merging them into the target branch. The GitHub Merge Queue, also known as the automatic merge queue, is a feature provided by GitHub that helps manage the order in which pull requests are merged into a repository's base branch.

https://github.com/orgs/community/discussions/58523#discussioncomment-6227114

**Reply by apoelstra** [reply]
Ok. I'll try turning it on the next time there's a simple merge candidate. I went through the Github settings and it was super unclear was was going on -- it said "merges can only be done through the merge queue" and had an option for how to do the merges, all of which corresponded to the green "merge" button and none of which were described as "you push a merge commit to master".

**Comment by tcharding** [comment]
I don't do the merging round here so my opinion isn't worth much but from my perspective the dev/review/merge process we have is pretty good. This has only happened twice (IIRC) in 3 years so its not worth putting too much time into. Also this repo is "special" in that @apoelstra and I basically know of (review) every single patch so we more or less know exactly what is going on. I'm not sure how this effects other contributors though.

**Reply by apoelstra** [reply]
>This has only happened twice (IIRC) in 3 years

If you mean "the build breaking because of merge conflicts", I have noticed this and refused to merge many more than 3 times. AFAIR every time the build broke it was because somebody else merged something.

It would be good if we could do this in a more public and reliable way.

**Reply by dpc** [reply]
This feature is absolutely terribly documented and half assed (can't actually merge groups of PRs), but once enabled it's quite seamless. 

One thing I'll add is that remember you can conditionally do or not stuff in the Merge Queue. Eg. in Fedimint we run our e2e test suite 5 times in the Merge Queue, just to make sure no flakiness slipped in: https://github.com/fedimint/fedimint/blob/b00fee0f7092b4c4ff13a27c5f4a3adc7b78f098/.github/workflows/ci-nix.yml#L203 . They still sometimes do, but it's much harder. :D

E.g. I image you could do the reverse and run some stuff only in the MQ, just to get much faster feedback in the PR.

**Comment by apoelstra** [comment]
I tried turning it on and it blocked my pushes, exactly as the terrible docs implied.

So I think this doesn't work at all.
